src/java/htsjdk/samtools/SAMBinaryTagAndValue.java

```diff
@@ -45,17 +50,46 @@
 
     /**
      * @param tag tagname (in binary form) for this attribute
-     * @param value value for this attribute (must be of a type that implements {@link Serializable} or else serialization will fail)
-     *              Cannot be null.
+     * @param value value for this attribute (must be of a type that implements {@link Serializable}
+     *              or else serialization will fail). Cannot be null.
      */
     public SAMBinaryTagAndValue(final short tag, final Object value) {
         if (null == value) {
             throw new IllegalArgumentException("SAMBinaryTagAndValue value may not be null");
         }
+        if (!isAllowedAttributeValue(value)) {
+            throw new IllegalArgumentException("Attribute type " + value.getClass() + " not supported. Tag: " +
+                    SAMTagUtil.getSingleton().makeStringTag(tag));
+        }
         this.tag = tag;
         this.value = value;
     }
 
+    // Inspect the proposed value to determine if it is an allowed value type,
+    // and if the value is in range.
+    protected static boolean isAllowedAttributeValue(final Object value) {
+        if (value instanceof String ||
```

Comment:
I noticed the order is different here than it was before.  Do we have any idea about the relative distribution of sam attribute types?  I assume the most common should be on top to short circuit them.  I don't know if the original order was carefully designed or just thrown together, but unless we know that this ordering is better I might keep the origianl.